pyquil/quilbase.py

```diff
@@ -189,104 +199,63 @@ def __init__(
-                _format_qubits_out(self.qubits),
-            )
+            return set(self._gate.qubits)
+
+    def get_qubit_indices(self) -> Set[int]:
+        return {qubit.as_fixed() for qubit in self._gate.qubits}
 
     def controlled(self, control_qubit: Union[QubitDesignator, Sequence[QubitDesignator]]) -> "Gate":
         """
         Add the CONTROLLED modifier to the gate with the given control qubit or Sequence of control
         qubits.
         """
-        control_qubit = control_qubit if isinstance(control_qubit, Sequence) else [control_qubit]
-        for qubit in control_qubit:
-            qubit = unpack_qubit(qubit)
-            self.modifiers.insert(0, "CONTROLLED")
-            self.qubits.insert(0, qubit)
+        if isinstance(control_qubit, Sequence):
+            for qubit in control_qubit:
+                self._gate = self._gate.controlled(convert_to_rs_qubit(qubit))
+        else:
+            self._gate = self._gate.controlled(convert_to_rs_qubit(control_qubit))
+
         return self
 
     def forked(self, fork_qubit: QubitDesignator, alt_params: List[ParameterDesignator]) -> "Gate":
         """
         Add the FORKED modifier to the gate with the given fork qubit and given additional
         parameters.
         """
-        if not isinstance(alt_params, list):
-            raise TypeError("Gate params must be a list")
-        if len(self.params) != len(alt_params):
-            raise ValueError("Expected {} parameters but received {}".format(len(self.params), len(alt_params)))
-
-        fork_qubit = unpack_qubit(fork_qubit)
-
-        self.modifiers.insert(0, "FORKED")
-        self.qubits.insert(0, fork_qubit)
-        self.params += alt_params
+        self._gate = self._gate.forked(convert_to_rs_qubit(fork_qubit), convert_to_rs_expressions(alt_params))
 
         return self
 
     def dagger(self) -> "Gate":
         """
         Add the DAGGER modifier to the gate.
         """
-        self.modifiers.insert(0, "DAGGER")
-
+        self._gate = self._gate.dagger()
         return self
 
     def __repr__(self) -> str:
         return "<Gate " + str(self) + ">"
 
     def __str__(self) -> str:
-        if self.params:
-            return "{}{}{} {}".format(
-                " ".join(self.modifiers) + " " if self.modifiers else "",
-                self.name,
-                _format_params(self.params),
-                _format_qubits_str(self.qubits),
-            )
-        else:
-            return "{}{} {}".format(
-                " ".join(self.modifiers) + " " if self.modifiers else "",
-                self.name,
-                _format_qubits_str(self.qubits),
-            )
+        return str(self._gate)
+
+    def out(self) -> str:
+        """
+        .. deprecated:: 4.0
+           This method defers to __str__ and will be removed in future versions of pyQuil. Calls to out() should be
+           replaced with str(gate).
+        """
+        return self.__str__()
```

Comment:
Great callout. I'll keep this in mind as I cook up the placeholder. I like the idea of two different routes - one that's forgiving and one that fails loudly.

I don't think we'll have this option, though - 

```
__str__() will succeed as it will fallback on a representation that doesn't need to be valid Quil.
```

I don't think we should make `quil-rs` generate invalid Quil, so it's really about choosing whether to fail or to just resolve placeholders before printing.

Reply:
We've decided to preserve the `out()` methods for now. They likely won't raise, but gives us the option if we ever need to expand the API in that way.